### DISCUSSION TITLE
[DRM] How to receive uppercase key character like '?', '@'?

### DISCUSSION THREAD

**Discussion by Ziqi-Yang** [raysan5/raylib#3773] [Q&A]
I find that shift and caps lock doesn't work in DRM, and my program only receive key character like '1' instead of '!', 'a' instead of 'A'. Currently, I do the shift convertion for alphabetical characters myself: (in Zig)
```zig
    const char = r.GetCharPressed();
    switch (char) {
        32...126 => {
            var c: u8 = @intCast(char);
            // TODO handle other key shift like '1' -> '!'
            if ((char >= 'a' and char <= 'z')
                    and (status.capsLockOn
                             or r.IsKeyDown(r.KEY_LEFT_SHIFT)
                             or r.IsKeyDown(r.KEY_RIGHT_SHIFT))) {
                c = c - 32;  // capitalize character
            }
            _ = self.push(c);
        },
        else => {}
    }
```
However, I cannot convert '1' to '!' since the people's keyboard may have different mappings. Is there a way to do this shift convertion? I see there is `xkbcommon` but I don't know if it suits since I may need to specify the region name like 'US' to get the uppercase key of the key I input ( my assumption). In my tty(using fish shell), I can directly input `!` and other characters.

**Comment by Ziqi-Yang** [comment]
I learnt that `xkbcommon` don't need to specify the region name. It requires `evdev` key code, which is `scancode + 8`. Raylib handles `scancode -> key` relationship, but it only support US keymap now (see [source code](https://github.com/raysan5/raylib/blob/7ec43022c177cbf00b27c9e9ab067bd6889957a4/src/platforms/rcore_drm.c#L145)). I may manually define the relationship for now, but I still look for a better solution. Also I doubt whether all users have the same shift relationship if they all use US keymap?

**Comment by LievenPetersen** [comment]
I haven't used the DRM backend, but in general Raylib has 2 different ways of polling the pressed keys: `GetKeyPressed` and `GetCharPressed`.
 - `GetKeyPressed` will return a keycode in US layout which could be incorrectly mapped, as you described.
 - `GetCharPressed` should return the actual letter that was pressed on the keyboard (But modifier-keys are not recognized as well).
 Maybe look into `GetCharPressed`. 
Hope this can help :)

**Reply by Ziqi-Yang** [reply]
My mistake. I haven't include `r.GetCharPressed();` into the code in my question. I have already used `GetCharPressed` function but it can only receive lowercase characters in DRM mode.
